From ff280e2753b84e0b46cf043ec92502ff2134ce53 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <rgnanase@qti.qualcomm.com>
Date: Fri, 15 Nov 2024 18:19:17 +0530
Subject: [PATCH] QSDK:CP:wifi:ath12k: Set WDS vdev param in station mode

When STA operating in WDS mode, 4addr qos NULL frame to be sent to Root
AP in order for AP to create VLAN. Expected sequence in set 4 addr WMI
peer param with AP address and then queue QOS NULL frame.

In some race condition, QOS NULL frame is scheduled by firmware even
before 4 addr param is set though host sends the cmd and frame in proper
order. In order to solve the issue, firmware proposed to send
WMI_VDEV_PARAM_WDS from host is sent to firmware for STA after vdev
create.

Patch-dependency: None
Patch-work: None

Signed-off-by: Ramya Gnanasekar <rgnanase@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1794,7 +1794,8 @@ void ath12k_mac_op_sta_set_4addr(struct
 	if (enabled && !ahsta->use_4addr_set) {
 		ahsta->ppe_vp_num = ahvif->dp_vif.ppe_vp_num;
 		ahsta->vlan_iface = ahvif->vlan_iface;
-		wiphy_work_queue(hw->wiphy, &ahsta->set_4addr_wk);
+		if (!ahvif->set_wds_vdev_param)
+			wiphy_work_queue(hw->wiphy, &ahsta->set_4addr_wk);
 		ahsta->use_4addr_set = true;
 	}
 }
@@ -14925,6 +14926,7 @@ int ath12k_mac_vdev_create(struct ath12k
 	struct ath12k_wmi_vdev_create_arg vdev_arg = {0};
 	struct ath12k_wmi_peer_create_arg peer_param = {0};
 	struct ieee80211_bss_conf *link_conf = NULL;
+	struct wireless_dev *wdev = ieee80211_vif_to_wdev(vif);
 	u32 param_id, param_value;
 	u16 nss;
 	int i;
@@ -15252,6 +15254,17 @@ int ath12k_mac_vdev_create(struct ath12k
 	ath12k_mac_ap_ps_recalc(ar);
 	ath12k_dp_vdev_tx_attach(ar, arvif);
 
+	if (vif->type == NL80211_IFTYPE_STATION &&
+	    (wdev && wdev->use_4addr)) {
+		ret = ath12k_wmi_vdev_set_param_cmd(arvif->ar, arvif->vdev_id,
+						    WMI_VDEV_PARAM_WDS, 1);
+		if (ret) {
+			ath12k_warn(ar->ab, "failed to set WDS vdev param: %d\n", ret);
+			goto err_vdev_del;
+		}
+		ahvif->set_wds_vdev_param = true;
+	}
+
 	return ret;
 
 err_peer_del:
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -716,6 +716,7 @@ struct ath12k_vif {
 	 */
 	struct ieee80211_chanctx_conf chanctx;
 	struct ath12k_reg_tpc_power_info reg_tpc_info;
+	bool set_wds_vdev_param;
 };
 
 struct ath12k_vif_iter {
